_posts/2023-06-23-assembly-versions.md

```diff
@@ -0,0 +1,113 @@
+
+![](/assets/assembly-version-overview.png)
+
+What do you expect the outcome of the following code to be:
+
+```csharp
+var demoClassType = typeof(DemoClass);
+Console.WriteLine(demoClassType.AssemblyQualifiedName); 
+// -> "MyLibrary.DemoClass, MyLibrary, Version=2.0.0.0, Culture=neutral, PublicKeyToken=null"
+
+var t1 = Type.GetType("MyLibrary.DemoClass, MyLibrary, Version=2.0.0.0, Culture=neutral, PublicKeyToken=null");
+Console.WriteLine("Type.GetType for version 2: " + t1);
+
+var t2 = Type.GetType("MyLibrary.DemoClass, MyLibrary, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null");
+Console.WriteLine("Type.GetType for version 1: " + t2);
+
+var t3 = Type.GetType("MyLibrary.DemoClass, MyLibrary, Version=3.0.0.0, Culture=neutral, PublicKeyToken=null");
+Console.WriteLine("Type.GetType for version 3: " + t3);
+```
+
+100 Points if your answer was the obligatory "it depends" ;) Let's look at a few cases:
+
+## .NET Framework
+
+Brining back ye olde .NET Framework for some nostalgia first. The [official documentation](https://learn.microsoft.com/en-us/troubleshoot/developer/visualstudio/general/assembly-version-assembly-file-version) states:
+
+>  At runtime, Common Language Runtime (CLR) looks for assembly with this version number to load. But remember this version is used along with name, public key token and culture information only if the assemblies are strong-named signed. If assemblies aren't strong-named signed, only file names are used for loading.
+
+Since I'm not using strong-naming here, the version should not be relevant. And indeed, all calls to `Type.GetType` return a result (which is equal to the result of `t1`).
+
+```
+MyLibrary.DemoClass, MyLibrary, Version=2.0.0.0, Culture=neutral, PublicKeyToken=null
+Type.GetType for version 2: MyLibrary.DemoClass
+Type.GetType for version 1: MyLibrary.DemoClass
+Type.GetType for version 3: MyLibrary.DemoClass
+```
+
+## .NET (Core)
+
+It hasn't been that easy to find some documentation on the behavior on .NET. But .NET changed a lot under the hood since AppDomains aren't really a thing anymore, so while things are kinda similar, sometimes they are not. The above code returns:
+
+```
+MyLibrary.DemoClass, MyLibrary, Version=2.0.0.0, Culture=neutral, PublicKeyToken=null
+Type.GetType for version 2: MyLibrary.DemoClass
+Type.GetType for version 1: MyLibrary.DemoClass
+Type.GetType for version 3:
+```
+
+So as long as we're loading the same or a lower version number of the type, it will be found. If we load a higher number however, .NET returns `null`.
```

Comment:
```suggestion
It will be found as long as we're loading the same or a lower version number of the type. However, if we try to load a higher number, .NET returns `null`.
```